src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerSessionMonitor.cs

```diff
@@ -186,113 +185,96 @@ private bool RemoveSession(int sessionId, out ProfilerSession session)
-                                SendEventsToListeners(viewerId, events, eventsLost);
-                            }
-                        }
-                    }
-                });
+                profileEvent.Values.Add(kvp.Key, kvp.Value.ToString());
             }
-        }
-
-        private List<ProfilerEvent> PollSession(ProfilerSession session)
-        {
-            var events = new List<ProfilerEvent>();
-            try
+            foreach (var kvp in xEvent.Actions)
             {
-                if (session == null || session.XEventSession == null)
-                {
-                    return events;
-                }
-
-                var targetXml = session.XEventSession.GetTargetXml();
+                profileEvent.Values.Add(kvp.Key, kvp.Value.ToString());
+            }
+            var eventList = new List<ProfilerEvent>();
+            eventList.Add(profileEvent);
+            var eventsLost = session.EventsLost;
 
-                XmlDocument xmlDoc = new XmlDocument();
-                xmlDoc.LoadXml(targetXml);
+            if (eventList.Count > 0 || eventsLost)
+            {
+                session.FilterOldEvents(eventList);
+                eventList = session.FilterProfilerEvents(eventList);
+                // notify all viewers of the event.
+                List<string> viewerIds = this.sessionViewers[session.XEventSession.Id];
 
-                var nodes = xmlDoc.DocumentElement.GetElementsByTagName("event");
-                foreach (XmlNode node in nodes)
+                foreach (string viewerId in viewerIds)
                 {
-                    var profilerEvent = ParseProfilerEvent(node);
-                    if (profilerEvent != null)
+                    if (allViewers[viewerId].active)
                     {
-                        events.Add(profilerEvent);
+                        SendEventsToListeners(viewerId, eventList, eventsLost);
                     }
                 }
             }
-            catch (XEventException)
-            {
-                SendStoppedSessionInfoToListeners(session.XEventSession.Id);
-                ProfilerSession tempSession;
-                RemoveSession(session.XEventSession.Id, out tempSession);
-            }
-            catch (Exception ex)
-            {
-                Logger.Write(TraceEventType.Warning, "Failed to poll session. error: " + ex.Message);
-            }
-            finally
+        }
+
+         /// <summary>
+        /// Nulls out properties in ConnectionDetails that aren't compatible with XElite.
+        /// </summary>
+        private ConnectionDetails CreateXEliteConnectionDetails(ConnectionDetails connDetails){
+            connDetails.ConnectRetryCount = null;
+            connDetails.ConnectRetryInterval = null;
+            connDetails.MultiSubnetFailover = null;
+            return connDetails;
+        }
+
+        /// <summary>
+        /// Creates a stream for new XEvents if the session hasn't streamed yet.
```

Comment:
I'm not seeing anywhere that you're checking if the session "hasn't streamed yet". Why do you need that check anyways?

Reply:
This code is only run when the session hasn't started streaming yet, this code should only be run once otherwise we would have multiple constantly running streams coming from the same session and it would lead to tons of duplicate events or possibly worse. That check is done outside of this code where ProcessStream is called.

Reply:
How/why would we get multiple streams attempting to be ran? It definitely seems like that if that was ever to happen we should be throwing an error, right?

Reply:
The ProcessStreams task/function constantly cycles through the available monitored sessions (this is a holdover when SMO was polling each session constantly), and checks each session whether they are already streaming or not before starting. I'm just saying that without the session stream check, it would try to create another stream using the same session.